src/targets/cpu/lowering.cpp

```diff
@@ -187,37 +187,61 @@ struct cpu_convolution
     {
         argument result{output_shape};
         result.visit([&](auto output) {
-            using type = typename decltype(output)::value_type;
             visit_all(args[0], args[1])([&](auto input, auto weights) {
-                auto in   = input.get_shape().lens();
-                auto in_h = in[2];
-                auto in_w = in[3];
-
-                auto wei   = weights.get_shape().lens();
-                auto wei_n = wei[0];
-                auto wei_c = wei[1];
-                auto wei_h = wei[2];
-                auto wei_w = wei[3];
-
-                par_dfor(output_shape.lens()[0],
-                         output_shape.lens()[1],
-                         output_shape.lens()[2],
-                         output_shape.lens()[3])(
-                    [&](std::size_t o, std::size_t w, std::size_t i, std::size_t j) {
-                        const auto start_x  = i * op.stride[0] - op.padding[0];
-                        const auto start_y  = j * op.stride[1] - op.padding[1];
-                        const auto group_id = w / (wei_n / op.group);
+                auto in_lens = input.get_shape().lens();
 
-                        type acc = type{0};
-                        dfor(wei_c, wei_h, wei_w)([&](std::size_t k, std::size_t x, std::size_t y) {
-                            const auto in_x  = start_x + x;
-                            const auto in_y  = start_y + y;
-                            const auto in_ch = group_id * wei_c + k;
-                            if(in_x >= 0 && in_x < in_h && in_y >= 0 && in_y < in_w)
-                                acc += input(o, in_ch, in_x, in_y) * weights(w, k, x, y);
-                        });
-                        output(o, w, i, j) = acc;
+                auto wei_lens                     = weights.get_shape().lens();
+                auto wei_n                        = wei_lens[0];
+                auto wei_c                        = wei_lens[1];
+                std::vector<std::size_t> win_size = {};
```

Comment:
would it be better to initialize `win_size` with the iterators from `wei_lens` instead?
like: `std::vector<std::size_t>win_size(wei_lens.begin() + 1, wei_lens.end());`

Reply:
Yea, let me update that.